primap2/pm2io/_data_reading.py

```diff
@@ -465,230 +526,161 @@ def read_wide_csv_file_if(
 
-    columns = read_data.columns.values
+    # TODO: add additional mapping functions here
+    mapping_functions = {
+        "PRIMAP1": {
+            "category": convert_ipcc_code_primap_to_primap2,
+            "entity": convert_entity_gwp_primap,
+        }
+    }
 
-    # TODO: add additional conversion functions here!
     meta_mapping_df = {}
-    if meta_mapping is not None:
-        # preprocess meta_mapping
-        for column in meta_mapping:
-            if isinstance(meta_mapping[column], str):
-                if column == "category":
-                    # print('creating meta_mapping dict for ' + column)
-                    values_to_map = read_data[coords_cols[column]].unique()
-                    if meta_mapping[column] == "PRIMAP1":
-                        # print("using PRIMAP converter")
-                        values_mapped = map(
-                            convert_ipcc_code_primap_to_primap2, values_to_map
-                        )
-                    else:
-                        logger.error(
-                            "Unknown metadata mapping "
-                            + meta_mapping[column]
-                            + " for column "
-                            + column
-                            + "."
-                        )
-                        raise ValueError(
-                            "Unknown metadata mapping "
-                            + meta_mapping[column]
-                            + " for column "
-                            + column
-                            + "."
-                        )
-                elif column == "entity":
-                    values_to_map = read_data[coords_cols[column]].unique()
-                    if meta_mapping[column] == "PRIMAP1":
-                        # print("using PRIMAP converter")
-                        values_mapped = list(
-                            map(convert_entity_gwp_primap, values_to_map)
-                        )
-                    else:
-                        logger.error(
-                            "Unknown metadata mapping "
-                            + meta_mapping[column]
-                            + " for column "
-                            + column
-                            + "."
-                        )
-                        raise ValueError(
-                            "Unknown metadata mapping "
-                            + meta_mapping[column]
-                            + " for column "
-                            + column
-                            + "."
-                        )
-                elif column == "unit":
-                    pass  # units are processed later
-                else:
+    # preprocess meta_mapping
+    for column, mapping in meta_mapping.items():
+        if column == "unit":  # handled in harmonize_units
+            continue
+
+        if isinstance(mapping, str) or callable(mapping):
```

Comment:
Hard coding the functions was actually done on purpose because I didn't like the idea that a configuration file can basically insert code into your program by defining arbitrary functions which are called. If I understand correctly you can pass user defined functions or classes with a __call__ function to the function, is that right? On the one hand that's neat, because datasets can come with their own mapping functions, on the other hand one has to take extra care (but if we use py files to define data reading we have the issue anyway) The code is definitely more elegant and it's easier to extend in this way.